fs/zbd_zenfs.cc

```diff
@@ -544,20 +579,33 @@ Zone *ZonedBlockDevice::AllocateZone(Env::WriteLifeTimeHint file_lifetime) {
 
     if (active_io_zones_.load() < max_nr_active_io_zones_) {
       for (const auto z : io_zones) {
-        if ((!z->open_for_write_) && z->IsEmpty()) {
-          z->lifetime_ = file_lifetime;
-          allocated_zone = z;
-          active_io_zones_++;
-          new_zone = 1;
-          break;
+        if (z->SetBusy()) {
```

Comment:
IIRC, SetBusy() would return true if it goes and set busy_ as true, else it returns false.
In case it does set it to true and if the zone is not empty(say it is a closed zone), then we will be setting this zone as busy and unset it in the else part below . 

Wouldn't it be better if we do a if ( ! IsBusy()) here and then actually do SetBusy() when the zone is allocated.

Reply:
`Zone::IsBusy()` has relaxed memory ordering, so none of the fields of `Zone` is guaranteed to be observed to the value they were set to, before the last owner of the zone called `Zone::UnsetBusy()`. That is to make `Zone::IsBusy()` faster.

We could change `Zone::IsBusy()` to stronger memory ordering, but we would still have a race on `Zone::lifetime_`.

Another approach could be to make a lockless implementation where the `Zone` pointer is atomically updated in-place in the io_zones list with a compare and swap operation. You would have to do this update every time you wanted to change a field of a `Zone`, and you would pass zones around as indexes to the list instead of as pointers. You would still need acquire/release ordering for the compare and swap operation though.

Could be interesting to benchmark the two approaches against each other.

Reply:
I think we should prioritize that the code should be a) robust and b) easy to understand. The first six rules of optimizations are: don't do it.

As long as Unset/SetBusy is safe to be accessed from different thtreads, i'm good.

Reply:
> I think we should prioritize that the code should be a) robust and b) easy to understand. 

> As long as Unset/SetBusy is safe to be accessed from different thtreads, i'm good.

I Agree. The current name of the methods does not really convey that you grab a mutex when you call them. How about renaming to `Acquire()` and `Release()` ?

Reply:
Sounds good